packages/sync/src/offline/OfflineStore.ts

```diff
@@ -7,39 +7,51 @@ import { OperationQueueEntry, OfflineItem } from "./OperationQueueEntry";
  */
 export class OfflineStore {
 
+  private offlineStorage: PersistentStore<OfflineItem>;
   private storage: PersistentStore<PersistedData>;
-  private readonly storageKey: string;
+  private arrayOfKeys: number[];
 
   constructor(clientConfig: DataSyncConfig) {
     this.storage = clientConfig.storage as PersistentStore<PersistedData>;
-    this.storageKey = clientConfig.mutationsQueueName;
+    this.offlineStorage = clientConfig.offlineStorage as PersistentStore<OfflineItem>;
+    this.arrayOfKeys = [];
   }
 
   /**
-   * Save data to store
+   * Save an entry to store
    *
-   * @param queue - array of offline elements to store
+   * @param entry - the entry to be saved
    */
-  public async persistOfflineData(queue: OperationQueueEntry[]) {
-    const offlineItems = queue.map((item: OperationQueueEntry) => {
-      return item.toOfflineItem();
-    });
-    if (this.storage && this.storageKey) {
-      await this.storage.setItem(this.storageKey, JSON.stringify(offlineItems));
-    }
+  public async saveEntry(entry: OperationQueueEntry) {
+    this.arrayOfKeys.push(entry.id);
+    this.storage.setItem("offlineKeys", JSON.stringify(this.arrayOfKeys));
+    await this.offlineStorage.setItem(entry.id.toString(), entry.toOfflineItem());
+  }
+
+  /**
+   * Remove an entry from the store
+   *
+   * @param queue - the entry to be removed
+   */
+  public async removeEntry(entry: OperationQueueEntry) {
+    this.arrayOfKeys.slice(this.arrayOfKeys.indexOf(entry.id));
+    this.storage.setItem("offlineKeys", JSON.stringify(this.arrayOfKeys));
+    await this.offlineStorage.removeItem(entry.id.toString());
   }
 
   /**
    * Fetch data from the
    */
   public async getOfflineData(): Promise<OfflineItem[]> {
-    const stored = this.storage.getItem(this.storageKey);
-    let offlineData;
-    if (stored) {
-      offlineData = JSON.parse(stored.toString());
-    } else {
-      offlineData = [];
+    const keys = await this.storage.getItem("offlineKeys");
+    const offlineItems: OfflineItem[] = [];
+    if (keys) {
+      this.arrayOfKeys = JSON.parse(keys.toString());
+      for (const key of this.arrayOfKeys) {
+        const item = await this.offlineStorage.getItem(key.toString());
```

Comment:
This breaks `window.localStorage` as it returns string etc.
It will be best to check if returned object is string or object and then apply stringify if needed.

Reply:
Same AsyncStorage from react: https://github.com/react-native-community/react-native-async-storage/blob/master/lib/AsyncStorage.js#L75